A74-xds-config-tears.md

```diff
@@ -0,0 +1,346 @@
+API will look like this (C++ syntax, various implementation details elided):
+
+```c++
+class Watcher {
+ public:
+  virtual ~Watcher() = default;
+
+  // Called when a new xDS config is available.
+  virtual void OnUpdate(XdsConfig config) = 0;
+
+  // These methods are invoked when there is an error or
+  // does-not-exist on LDS or RDS only.  The context will be a
+  // human-readable string indicating the scope in which the error
+  // occurred (e.g., the resource type and name).
+  virtual void OnError(absl::string_view context, absl::Status status) = 0;
+  virtual void OnResourceDoesNotExist(std::string context) = 0;
+};
+```
+
+The xds resolver will implement this watcher interface and pass the
+watcher object to the XdsDependencyManager at instantiation time.
+
+The updates delivered to this watcher will contain a complete xDS
+configuration, including all necessary resources.  The xDS configuration
+will look something like this (C++ syntax, various implementation
+details elided):
+
+```c++
+struct XdsConfig {
+  // Listener resource.
+  XdsListenerResource listener;
+  // RouteConfig resource.  Will be populated even if RouteConfig is
+  // inlined into the Listener resource.
+  XdsRouteConfigResource route_config;
+  // Virtual host.  Points into route_config.  Will always be non-null.
+  XdsRouteConfigResource::VirtualHost* virtual_host;
+
+  struct ClusterConfig {
+    // Cluster resource.
+    XdsClusterResource cluster;
+
+    // Endpoint info for EDS and LOGICAL_DNS clusters.  If there was an
+    // error, endpoints will be null and resolution_note will be set.
+    struct EndpointConfig {
+      XdsEndpointResource endpoints;
+      std::string resolution_note;
+    };
+    // The list of leaf clusters for an aggregate cluster.
+    struct AggregateConfig {
+      std::vector<absl::string_view> leaf_clusters;
+    };
+    std::variant<EndpointConfig, AggregateConfig> children;
+  };
+  // Cluster map.  A cluster will have a non-OK status if either
+  // (a) there was an error and we did not already have a valid
+  // resource or (b) the resource does not exist.
+  std::map<std::string, absl::StatusOr<ClusterConfig>> clusters;
+};
+```
+
+Note that the watcher's OnError() and OnResourceDoesNotExist()
+methods will be invoked only for problems with the listener and route
+configuration resources.  For cluster-specific errors, we need to
+retain the existing behavior, which is that it causes a problem for the
+individual cluster but does not affect traffic sent to any other cluster.
+Therefore, when the XdsDependencyManager sees an error on a cluster for
+which it did not previously have data, or if it sees a does-not-exist
+notification on a cluster, it will report that by returning a non-OK
+status for the cluster in the XdsConfig.  Similarly, if there is a problem
+obtaining endpoint data from EDS or DNS, the XdsDependencyManager will
+report that problem via the cluster's resolution_note field.
```

Comment:
Why not also make the `EndpointConfig` contain a `StatusOr`?

Reply:
Because there are cases where we set both the `endpoints` and the `resolution_note` field, such as when the EDS update contains an empty locality.

Note that this structure isn't really new; it just models the existing mechanism for passing endpoint addresses from the resolver to the LB policy tree, and we use the same representation for both.